File: packages/chain/verifier/mnemaverify/schema.py
"""Section 4.1 - the field declarations, read from the artifact that publishes them.

GAP G08, CLOSED. Section 4 has always promised that "a reader rebuilds the event from the
fields its kind declares and rejects any other". Those declarations used to be published
NOWHERE: the canonical vectors give one exemplar per kind, and from an exemplar a required
field and an optional one that happens to be present look identical. So this reader had
only byte identity - which catches a field added to a line that was already written, and
CANNOT catch a newly appended event carrying one. It was measured accepting exactly that,
which the product refuses, and which a party with no key at all can produce.

`event-schema.json` is the artifact that closed it, and this module is section 4.1
implemented against it. Nothing here knows any kind by name: every rule comes out of the
file, and a kind, a field or a rule the file grows is a kind, a field or a rule this reader
applies without being edited.

WHAT THIS MODULE DELIBERATELY DOES NOT DO: trust the file's own `rules` glossary. The file
carries a one-line gloss of each rule so it is readable on its own, and section 4.1 states
the same vocabulary normatively. THE PROSE IS WHAT IS IMPLEMENTED HERE. What the glossary
is used for is the opposite check - that the file does not use a rule name it does not
define, and that it does not use one this reader has never heard of. A rule this reader
cannot apply is reported and the line it applies to is UNCHECKED, never quietly accepted:
"I do not know this rule" and "this line is fine" are not the same answer.
"""

from __future__ import annotations

import json
import os
import re
from typing import Any, NamedTuple

from .canonical import strict_loads
from .verdict import Refusal

DEFAULT_SCHEMA = os.path.normpath(
    os.path.join(os.path.dirname(os.path.abspath(__file__)), "..", "..", "event-schema.json")
)

# Section 4.1's `instant`: the exact spelling `toISOString` produces. The shape is a regex
# and the VALUE is a round trip - a pattern alone admits month 13 and day 32, which are not
# instants any clock produced.
_INSTANT = re.compile(r"^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\.\d{3}Z$")

# Every rule name section 4.1 defines. A file naming one that is not here is a file this
# reader is too old for, and it says so rather than skipping the field.
KNOWN_RULES = frozenset(
    {
        "string",
        "string?",
        "string|null",
        "boolean",
        "count",
        "fields?",
        "string[]?",
        "version",
        "kind",
        "instant",
    }
)

# The largest whole number a double carries exactly. Section 4.1 says `count` is "a whole
# number of at least 1 (a safe integer)"; anything past this is a value that cannot survive
# the round trip through the JSON number section 1 specifies.
_MAX_SAFE_INTEGER = 2**53 - 1


class Contract(NamedTuple):
    kind: str
    v: int
    payload: dict[str, str]


class Schema(NamedTuple):
    """The published declarations, as this reader uses them."""

    envelope: dict[str, str]
    transition_fields: dict[str, str]
    contracts: dict[tuple[str, int], Contract]
    rules: dict[str, str]
    path: str
    unknown_rules: list[str]

    def top_level_keys(self) -> frozenset[str]:
        """Section 4.1: the keys of `envelope`, plus `payload`."""
        return frozenset(self.envelope) | {"payload"}


def load(path: str = DEFAULT_SCHEMA) -> Schema:
    """Read the artifact. Raises Refusal when it is not the file section 4.1 describes."""
    try:
        with open(path, "rb") as handle:
            document = strict_loads(handle.read().decode("utf-8"))
    except OSError as exc:
        raise Refusal("4", f"could not read the field declarations at {path}: {exc}") from exc
    if not isinstance(document, dict):
        raise Refusal("4", f"{path} is not the published schema")
    for key in ("envelope", "transitionFields", "contracts"):
        if key not in document:
            raise Refusal("4", f"{path} has no {key}, so it is not the published schema")

    envelope = _rule_map(document["envelope"], "envelope", path)
    transition = _rule_map(document["transitionFields"], "transitionFields", path)
    rows = document["contracts"]
    if not isinstance(rows, list) or not rows:
        raise Refusal("4", f"{path} declares no contracts")
    contracts: dict[tuple[str, int], Contract] = {}
    for row in rows:
        if not isinstance(row, dict) or not isinstance(row.get("kind"), str):
            raise Refusal("4", f"{path} holds a contract with no kind")
        v = row.get("v")
        if not isinstance(v, int) or isinstance(v, bool):
            raise Refusal("4", f"{path}: contract {row['kind']} has no integer v")
        payload = _rule_map(row.get("payload", {}), f"contract {row['kind']}@{v}", path)
        key = (row["kind"], v)
        if key in contracts:
            raise Refusal("4", f"{path} declares {row['kind']}@{v} twice")
        contracts[key] = Contract(row["kind"], v, payload)

    glossary = document.get("rules")
    rules = glossary if isinstance(glossary, dict) else {}
    used = set(envelope.values()) | set(transition.values())
    for contract in contracts.values():
        used |= set(contract.payload.values())
    unknown = sorted(used - KNOWN_RULES)
    return Schema(envelope, transition, contracts, rules, path, unknown)


def _rule_map(value: Any, where: str, path: str) -> dict[str, str]:
    if not isinstance(value, dict):
        raise Refusal("4", f"{path}: {where} is not an object of field rules")
    for field, rule in value.items():
        if not isinstance(rule, str):
            raise Refusal("4", f"{path}: {where}.{field} is not a rule name")
    return dict(value)


def undefined_rules(schema: Schema) -> list[str]:
    """Rule names the file uses and its own glossary does not define.

    Not a refusal of a record - it is a defect of the ARTIFACT, and it is reported as one.
    An artifact that names a rule it does not gloss is one a stranger has to guess at,
    which is the failure the whole file exists to end.
    """
    if not schema.rules:
        return []
    used: set[str] = set(schema.envelope.values()) | set(schema.transition_fields.values())
    for contract in schema.contracts.values():
        used |= set(contract.payload.values())
    return sorted(used - set(schema.rules))


class Rebuilt(NamedTuple):
    """One event, rebuilt from exactly what its contract declares."""

    event: dict[str, Any]
    contract: Contract


def rebuild(schema: Schema, event: Any) -> Rebuilt:
    """Section 4 and 4.1: rebuild the event from its declared fields, refusing any other.

    The rebuild is returned rather than a boolean, because "the fields it declares" is only
    half of section 4: the other half is that the rebuilt event canonicalizes to the stored
    bytes, and a caller that cannot see the rebuild cannot check that.
    """
    if not isinstance(event, dict):
        raise Refusal("4", "an event that is not a JSON object")
    kind = event.get("kind")
    if not isinstance(kind, str) or not kind:
        raise Refusal("4.1", "an event with no `kind`, so no contract selects it")
    v = event.get("v")
    if not isinstance(v, int) or isinstance(v, bool) or v < 1:
        raise Refusal("4.1", f"event {kind!r} carries no whole-number `v` of at least 1")
    contract = schema.contracts.get((kind, v))
    if contract is None:
        raise Refusal(
            "4.1",
            f"no published contract declares {kind}@{v}: a reader refuses a pair the table "
            "does not declare rather than guessing at the fields",
        )

    declared_top = schema.top_level_keys()
    extra = sorted(set(event) - declared_top)
    if extra:
        raise Refusal("4.1", f"event {kind!r} carries top-level field(s) {extra}, which nothing declares")

    rebuilt: dict[str, Any] = {}
    for field, rule in schema.envelope.items():
        value = _apply(kind, field, rule, event.get(field, _MISSING))
        if value is not _MISSING:
            rebuilt[field] = value

    payload = event.get("payload", _MISSING)
    if not isinstance(payload, dict):
        raise Refusal("4.1", f"event {kind!r} needs an object payload")
    forged = sorted(set(payload) - set(contract.payload))
    if forged:
        raise Refusal(
            "4.1",
            f"event {kind!r} carries payload field(s) {forged}, which its contract does not "
            "declare, so they would ride along into the signed bytes",
        )
    rebuilt_payload: dict[str, Any] = {}
    for field, rule in contract.payload.items():
        value = _apply(kind, f"payload.{field}", rule, payload.get(field, _MISSING), schema)
        if value is not _MISSING:
            rebuilt_payload[field] = value
    rebuilt["payload"] = rebuilt_payload
    return Rebuilt(rebuilt, contract)


class _Missing:
    """The absence of a key, which is not the same value as any key could hold."""

    def __repr__(self) -> str:  # pragma: no cover - a debugging convenience
        return "<absent>"


_MISSING = _Missing()


def _apply(kind: str, field: str, rule: str, value: Any, schema: Schema | None = None) -> Any:
    """Section 4.1's vocabulary, one rule at a time. Returns _MISSING for a legitimate absence."""
    if rule == "string":
        return _require_string(kind, field, value)
    if rule == "string?":
        return _MISSING if value is _MISSING else _require_string(kind, field, value)
    if rule == "string|null":
        if value is None:
            return None
        return _require_string(kind, field, value)
    if rule == "boolean":
        # Nothing is coerced: `"false"`, `0` and `null` are all falsy in the language this
        # format is written in and none of them is the position of a switch.
        if not isinstance(value, bool):
            raise Refusal("4.1", f"event {kind!r} needs true or false at {field}")
        return value
    if rule == "count":
        if isinstance(value, bool) or not isinstance(value, int) or value < 1:
            raise Refusal("4.1", f"event {kind!r} needs a whole number of at least 1 at {field}")
        if value > _MAX_SAFE_INTEGER:
            raise Refusal("4.1", f"event {kind!r} has a count past the safe integers at {field}")
        return value
    if rule == "string[]?":
        if value is _MISSING:
            return _MISSING
        if not isinstance(value, list) or not value:
            raise Refusal("4.1", f"event {kind!r} needs a non-empty array at {field}")
        for at, item in enumerate(value):
            if not isinstance(item, str) or not item:
                raise Refusal("4.1", f"event {kind!r} needs a non-empty string at {field}[{at}]")
        return list(value)
    if rule == "fields?":
        if value is _MISSING:
            return _MISSING
        if schema is None:  # pragma: no cover - only the payload pass carries `fields?`
            raise Refusal("4.1", f"event {kind!r} declares {field} outside a payload")
        return _rebuild_fields(kind, field, value, schema)
    if rule == "version":
        if isinstance(value, bool) or not isinstance(value, int) or value < 1:
            raise Refusal("4.1", f"event {kind!r} needs a whole number of at least 1 at {field}")
        return value
    if rule == "kind":
        return _require_string(kind, field, value)
    if rule == "instant":
        return _require_instant(kind, field, value)
    raise Refusal("4.1", f"event {kind!r} declares {field} under rule {rule!r}, which this reader does not know")


def _rebuild_fields(kind: str, field: str, value: Any, schema: Schema) -> dict[str, Any]:
    if not isinstance(value, dict):
        raise Refusal("4.1", f"event {kind!r} needs an object at {field}")
    forged = sorted(set(value) - set(schema.transition_fields))
    if forged:
        raise Refusal("4.1", f"event {kind!r} carries {field} key(s) {forged}, which nothing declares")
    rebuilt: dict[str, Any] = {}
    for name, rule in schema.transition_fields.items():
        got = _apply(kind, f"{field}.{name}", rule, value.get(name, _MISSING), schema)
        if got is not _MISSING:
            rebuilt[name] = got
    if not rebuilt:
        # Section 4.1: never empty. An empty `fields` carries no proof, and admitting it
        # would make `{}` a second, byte-distinct spelling of "no fields".
        raise Refusal("4.1", f"event {kind!r} has an empty {field}; the declaration says omit it")
    return rebuilt


def _require_string(kind: str, field: str, value: Any) -> str:
    if not isinstance(value, str) or not value:
        raise Refusal("4.1", f"event {kind!r} needs a non-empty string at {field}")
    return value


def _require_instant(kind: str, field: str, value: Any) -> str:
    text = _require_string(kind, field, value)
    if _INSTANT.match(text) is None or not _is_a_real_date(text):
        raise Refusal(
            "4.1",
            f"event {kind!r} needs a UTC millisecond instant at {field}, and carries {text!r}",
        )
    return text


def _is_a_real_date(text: str) -> bool:
    """The round trip the shape alone does not give: month 13 and day 32 match the pattern."""
    from datetime import datetime, timezone

    try:
        parsed = datetime(
            int(text[0:4]),
            int(text[5:7]),
            int(text[8:10]),
            int(text[11:13]),
            int(text[14:16]),
            int(text[17:19]),
            int(text[20:23]) * 1000,
            tzinfo=timezone.utc,
        )
    except ValueError:
        return False
    return parsed.strftime("%Y-%m-%dT%H:%M:%S.") + f"{parsed.microsecond // 1000:03d}Z" == text

File: packages/chain/src/events/schema.ts
/**
 * The field declarations, as DATA — the one place a kind says which fields it
 * carries, and the only thing the reader rebuilds an event from.
 *
 * `FORMAT.md` section 4 has always promised that "a reader rebuilds the event
 * from the fields its kind declares and rejects any other". Until this module
 * existed, those declarations were a `switch` of twenty arms inside `parse.ts`
 * and a key list beside it — readable by this codebase and by nothing else. An
 * independent verifier written from the document measured the consequence: an
 * event APPENDED above the last checkpoint with a field no kind declares was
 * read as verified by a stranger and as unreadable by this product. The entry
 * hash takes no key, so whoever can write the repository computes it; no
 * signature covers that window; and the envelope keys are all present. The only
 * thing that could have refused it was the declaration, and the declaration was
 * not published.
 *
 * So the declarations became data, and the data is published:
 * `event-schema.json` beside `FORMAT.md`, described by section 4.1.
 *
 * ONE SOURCE, AND WHY IT HAD TO BE THIS ONE. A hand-written schema beside a
 * hand-written reader is two places where the fields live, and two places drift
 * — silently, because the day they disagree is the day the second reader starts
 * accepting what this one refuses again, with nothing red. So the table below is
 * not a description of the reader: it IS the reader. `parse.ts` walks these rows
 * and has no per-kind branch left, and the published artifact is this table
 * serialized. A field that is not here is a field nothing validates, rebuilds,
 * or publishes.
 *
 * THE COMPILER HOLDS THE TOTALITY, and it holds it here in `src` rather than in
 * a test: `tsc -b` excludes `src/**\/*.test.ts` and vitest strips types without
 * checking them, so a mapped type declared in a test file is vacuous by
 * construction. {@link PAYLOAD_SCHEMA} is keyed by {@link EventKind}, so a new
 * kind does not compile until it has a row; each row is keyed by the payload's
 * own field names, so a field added to, renamed in, or removed from the catalog
 * does not compile until the row follows. That is stronger than the key list it
 * replaces, which was keyed by kind alone and let a field appear in an arm and
 * nowhere else.
 *
 * WHAT A RULE IS. A rule is one of a small closed vocabulary
 * ({@link FieldRule}), spelled the same way in this table and in the published
 * artifact, because a stranger has to be able to read it. The vocabulary is
 * small on purpose: every payload field of the catalog falls into six shapes,
 * and a seventh would be a decision somebody makes rather than a convenience
 * somebody reaches for.
 */

import type { CatalogEvent, EventKind, TransitionFields } from './catalog.js';
import type { Envelope } from './envelope.js';

/**
 * The vocabulary a payload field rule is spelled in.
 *
 * Each name is a refusal as much as a shape — the point of a closed catalog is
 * that a value which is ALMOST the right shape is a line two readers could
 * disagree about, so each rule below says exactly one thing and coerces nothing:
 *
 * - `string` — present, and a non-empty string.
 * - `string?` — absent, or a non-empty string. Absent is not `null` and not `""`.
 * - `string|null` — present, and either a non-empty string or an explicit `null`.
 *   The `null` is a VALUED absence (the state a birth transition left behind),
 *   which is why it is a spelling of its own rather than an optional.
 * - `boolean` — present, and exactly `true` or `false`. Nothing is coerced:
 *   `"false"`, `0` and `null` are all falsy in this language and none of them is
 *   the position of a switch.
 * - `count` — present, and a whole number of at least one. Every way a number can
 *   be almost-a-number (a float, a `-0`, a `1e3`, a NaN) is a way two readers
 *   could disagree about the same line.
 * - `fields?` — absent, or the transition-proof object declared by
 *   {@link TRANSITION_FIELDS_SCHEMA}, closed the same way and never empty.
 * - `string[]?` — absent, or a non-empty array of non-empty strings.
 */
export type FieldRule =
  | 'string'
  | 'string?'
  | 'string|null'
  | 'boolean'
  | 'count'
  | 'fields?'
  | 'string[]?';

/**
 * The vocabulary an ENVELOPE field rule is spelled in: the payload vocabulary
 * plus the two shapes only the envelope has.
 *
 * - `version` — a whole number of at least one, selecting the payload contract
 *   together with `kind`.
 * - `kind` — the discriminator itself: a non-empty string naming a declared kind.
 * - `instant` — the exact spelling `Date.prototype.toISOString` produces (UTC,
 *   millisecond precision, trailing `Z`) AND a real date. Every producer stamps
 *   `at` through the clock, which IS `toISOString`, so a well-formed `at` is not
 *   merely "some ISO-8601 string" but this one canonical spelling — which is what
 *   makes the ordering invariant enforceable rather than merely documented.
 */
export type EnvelopeRule = FieldRule | 'version' | 'kind' | 'instant';

/**
 * The envelope every kind carries, and the rule on each field.
 *
 * Keyed by `keyof Required<Envelope>`, so a field added to the envelope does not
 * compile until it is declared here — and therefore until it is published.
 *
 * THE ORDER HERE IS OBSERVABLE, which is not obvious and was measured. Canonical
 * bytes sort their keys (§1), so the order cannot move a digest — but the rebuilt
 * event is what a surface prints, and `JSON.stringify` keeps insertion order. So
 * this is the order an agent reading `--json` sees, and it is pinned by
 * `packages/code/src/cli.golden.test.ts`. It is the six required fields, in the
 * order the reader validates them, and then the two optional ones — which is both
 * what the product has always emitted and the more legible half-and-half for
 * somebody reading the published file.
 */
export const ENVELOPE_SCHEMA: { readonly [F in keyof Required<Envelope>]: EnvelopeRule } = {
  v: 'version',
  kind: 'kind',
  at: 'instant',
  who: 'string',
  signerFp: 'string',
  subject: 'string',
  which: 'string?',
  run: 'string?',
};

/**
 * The proof/context object a transition may carry.
 *
 * Every field is optional and the object is never empty: an empty `fields` carries
 * no proof, so treating it as absence is what stops it becoming a second,
 * byte-distinct spelling of "no fields".
 */
export const TRANSITION_FIELDS_SCHEMA: {
  readonly [F in keyof Required<TransitionFields>]: FieldRule;
} = {
  reason: 'string?',
  note: 'string?',
  feedback: 'string?',
  pr_url: 'string?',
  links: 'string[]?',
};

/** The payload contract of one kind: every declared field, with its rule. */
type PayloadSchemaOf<K extends EventKind> = {
  readonly [F in keyof Required<Extract<CatalogEvent, { kind: K }>['payload']>]: FieldRule;
};

/**
 * What each kind declares, and nothing else may ride along.
 *
 * The row order is the catalog's own kind order, and the field order within a row
 * is the order the reader validates in — which is what a reader's first refusal
 * names when a line is wrong in more than one way.
 */
export const PAYLOAD_SCHEMA: { readonly [K in EventKind]: PayloadSchemaOf<K> } = {
  'run.started': { agent: 'string', goal: 'string?' },
  'run.ended': { outcome: 'string?' },
  'task.created': { title: 'string' },
  'task.transitioned': {
    from: 'string|null',
    to: 'string',
    action: 'string',
    fields: 'fields?',
  },
  // `alternatives` is declared OPTIONAL, and that is a fact about the past rather
  // than a convenience: a decision recorded before the field existed omits it,
  // parses, and canonicalizes to exactly the bytes it was signed as. A schema that
  // published one exemplar per kind could never say this — which is the half of a
  // declaration a stranger cannot infer from an example.
  'decision.recorded': {
    title: 'string',
    rationale: 'string',
    adr: 'string',
    alternatives: 'string?',
  },
  'decision.transitioned': {
    from: 'string|null',
    to: 'string',
    action: 'string',
    by: 'string?',
    fields: 'fields?',
  },
  'identity.founded': { foundingFp: 'string' },
  'key.enrolled': { newFp: 'string', reverseSig: 'string' },
  'key.revoked': { revokedFp: 'string', reason: 'string' },
  'memory.captured': { content: 'string' },
  'observation.recorded': { about: 'string', topic: 'string', text: 'string' },
  'handoff.recorded': { fromAgent: 'string', toAgent: 'string' },
  // `rel` is an open literal string: any non-empty string is accepted, never
  // matched against a closed set, so a new relation label needs no upcaster and a
  // past link with an unfamiliar one is never rejected on read.
  'knowledge.linked': { target: 'string', rel: 'string' },
  'skill.created': { name: 'string', body: 'string' },
  // Mirrors task.transitioned: `from` is a state or the birth's null, and there is
  // no `by` — a skill is not relational.
  'skill.transitioned': {
    from: 'string|null',
    to: 'string',
    action: 'string',
    fields: 'fields?',
  },
  // No payload field at all: a consultation is entirely envelope. The empty row is
  // what makes ANY payload key on this kind a rejected line.
  'skill.consulted': {},
  'tail.pruned': {
    tail: 'string',
    throughHash: 'string',
    // The catalog's first numeric payload field, and the rule on it is the reader's
    // alone: a count of events in a tail that no longer exists is a claim nothing on
    // disk can be compared against once the cut has happened, so what CAN be said is
    // that it is a whole number of at least one. Zero is refused because a tail with
    // no events has no head to name either.
    // WHAT IS DELIBERATELY NOT DECLARED: whether the named tail is on disk. That is
    // the WRITE-side rule (see `unprovenWaiverReason`), and it must never become a
    // read-side one — the waiver exists to survive the cut, so the moment the tail it
    // names is gone, a reader applying the write rule would refuse the waiver, and
    // refusing one line refuses the whole tail it lives on, permanently.
    eventCount: 'count',
    reason: 'string',
  },
  // The catalog's only BOOLEAN, and the rule is that it is one: a `"off"`, a `0` or
  // a missing key are three ways for two readers to disagree about whether a channel
  // was on, and this kind is read to decide whether the product says anything at all.
  'channel.switched': { on: 'boolean', reason: 'string?' },
  // No payload field at all, the same as a consultation: the fact is entirely envelope.
  'channel.served': {},
  // Both are REQUIRED, and the rule especially: this is the one kind of the catalog
  // whose whole standing is that it names what caused it, so a charge with an absent
  // or empty citation is a line this reader refuses rather than one it lifts.
  'channel.asked': { rule: 'string', path: 'string' },
};

/**
 * What each rule name MEANS, in one line, carried inside the published artifact.
 *
 * The artifact is for a reader who has only the artifact and the document, so it
 * carries its own glossary rather than only rule names. The normative text is
 * `FORMAT.md` section 4.1 — this is the same sentence, travelling with the data,
 * so a file that arrives on its own is still readable. It is derived from one
 * table here for the same reason everything else in this module is: a glossary
 * that could disagree with the vocabulary is a second place the rules live.
 */
export const RULE_GLOSSARY: { readonly [R in EnvelopeRule]: string } = {
  string: 'present, and a non-empty string',
  'string?': 'absent, or a non-empty string; absent is not null and not the empty string',
  'string|null': 'present, and either a non-empty string or an explicit null',
  boolean: 'present, and exactly true or false; nothing is coerced',
  count: 'present, and a whole number of at least 1',
  'fields?': 'absent, or the transitionFields object below, closed the same way and never empty',
  'string[]?': 'absent, or a non-empty array of non-empty strings',
  version: 'present, and a whole number of at least 1; with kind it selects this contract',
  kind: 'present, and the non-empty string naming this contract',
  instant:
    'present, and the exact spelling Date.prototype.toISOString produces ' +
    '(UTC, millisecond precision, trailing Z) of a real date',
};

/** One published contract: the payload a `(kind, v)` pair declares. */
export interface PublishedContract {
  readonly kind: string;
  readonly v: number;
  readonly payload: { readonly [field: string]: FieldRule };
}

/** The published artifact, as data. */
export interface PublishedSchema {
  readonly schemaVersion: number;
  readonly describedBy: string;
  readonly rules: { readonly [rule: string]: string };
  readonly envelope: { readonly [field: string]: EnvelopeRule };
  readonly transitionFields: { readonly [field: string]: FieldRule };
  readonly contracts: readonly PublishedContract[];
}

/**
 * The published schema, built from the tables above and from nothing else.
 *
 * `kind` and `v` TOGETHER select one contract, which is the pair section 7
 * already made every event carry and the pair the interoperability world settles
 * this with (in-toto's `predicateType` is a URI carrying the major version, and
 * it changes when the change is incompatible). A contract list rather than a map
 * keyed by kind, because that is the shape a second version grows into: one more
 * row, not a reshaped file.
 *
 * The rows are sorted by `(kind, v)` so the file has one order, and the order is
 * not the declaration order of a table somebody may reorder.
 */
export function publishedSchema(latest: { readonly [kind: string]: number }): PublishedSchema {
  const contracts = Object.entries(PAYLOAD_SCHEMA)
    .map(([kind, payload]) => ({
      kind,
      v: latest[kind] ?? 1,
      payload: payload as { readonly [field: string]: FieldRule },
    }))
    .sort((a, b) => (a.kind === b.kind ? a.v - b.v : a.kind < b.kind ? -1 : 1));
  return {
    schemaVersion: 1,
    describedBy: 'FORMAT.md',
    rules: RULE_GLOSSARY,
    envelope: ENVELOPE_SCHEMA,
    transitionFields: TRANSITION_FIELDS_SCHEMA,
    contracts,
  };
}

/**
 * Where the published artifact is, resolved from this module.
 *
 * ONE site for the path, the same discipline `vectors.ts` states for the other
 * artifact: the test that checks the file, the guard that checks `FORMAT.md`, and
 * the sentence in `FORMAT.md` that tells a stranger where to find it all have to
 * name the same file. The depth is the same from `src/events/` and from
 * `dist/events/`, so it resolves in the suite and in the built package.
 */
export const SCHEMA_FILE_NAME = 'event-schema.json';
export const SCHEMA_FILE = new URL(`../../${SCHEMA_FILE_NAME}`, import.meta.url);

/**
 * The artifact's bytes, exactly as the file holds them: two-space JSON with a
 * trailing newline.
 *
 * Written out here rather than left to a test, because the file and the table are
 * checked BYTE for byte — a serializer that disagreed with the file's formatting
 * would make the guard fail for a reason that is not drift, and somebody would
 * "fix" it by loosening the comparison.
 */
export function publishedSchemaText(latest: { readonly [kind: string]: number }): string {
  return `${JSON.stringify(publishedSchema(latest), null, 2)}\n`;
}
